gio/src/subclass/action.rs

```diff
@@ -0,0 +1,254 @@
+            let ret = func(self.obj().unsafe_cast_ref::<Action>().to_glib_none().0);
+            ret != 0
+        }
+    }
+
+    fn parent_state(&self) -> Option<Variant> {
+        unsafe {
+            let type_data = Self::type_data();
+            let parent_iface =
+                type_data.as_ref().parent_interface::<Action>() as *const ffi::GActionInterface;
+
+            let func = (*parent_iface)
+                .get_state
+                .expect("no parent \"get_state\" implementation");
+            let ret = func(self.obj().unsafe_cast_ref::<Action>().to_glib_none().0);
+            from_glib_none(ret)
+        }
+    }
+
+    fn parent_change_state(&self, value: Variant) {
+        unsafe {
+            let type_data = Self::type_data();
+            let parent_iface =
+                type_data.as_ref().parent_interface::<Action>() as *const ffi::GActionInterface;
+
+            let func = (*parent_iface)
+                .change_state
+                .expect("no parent \"change_state\" implementation");
+            func(
+                self.obj().unsafe_cast_ref::<Action>().to_glib_none().0,
+                value.to_glib_none().0,
+            );
+        }
+    }
+
+    fn parent_activate(&self, parameter: Option<Variant>) {
+        unsafe {
+            let type_data = Self::type_data();
+            let parent_iface =
+                type_data.as_ref().parent_interface::<Action>() as *const ffi::GActionInterface;
+
+            let func = (*parent_iface)
+                .activate
+                .expect("no parent \"activate\" implementation");
+            func(
+                self.obj().unsafe_cast_ref::<Action>().to_glib_none().0,
+                parameter.to_glib_none().0,
+            );
+        }
+    }
+}
+
+impl<T: ActionImpl> ActionImplExt for T {}
+
+unsafe impl<T: ActionImpl> IsImplementable<T> for Action {
+    fn interface_init(iface: &mut glib::Interface<Self>) {
+        let iface = iface.as_mut();
+
+        iface.get_name = Some(action_get_name::<T>);
+        iface.get_parameter_type = Some(action_get_parameter_type::<T>);
+        iface.get_state_type = Some(action_get_state_type::<T>);
+        iface.get_state_hint = Some(action_get_state_hint::<T>);
+        iface.get_enabled = Some(action_get_enabled::<T>);
+        iface.get_state = Some(action_get_state::<T>);
+        iface.change_state = Some(action_change_state::<T>);
+        iface.activate = Some(action_activate::<T>);
+    }
+}
+
+unsafe extern "C" fn action_get_name<T: ActionImpl>(
+    actionptr: *mut ffi::GAction,
+) -> *const libc::c_char {
+    let instance = &*(actionptr as *mut T::Instance);
+    let imp = instance.imp();
+
+    imp.name().to_glib_none().0
+}
+
+unsafe extern "C" fn action_get_parameter_type<T: ActionImpl>(
+    actionptr: *mut ffi::GAction,
+) -> *const GVariantType {
+    let instance = &*(actionptr as *mut T::Instance);
+    let imp = instance.imp();
+
+    imp.parameter_type().to_glib_none().0
```

Comment:
Are all the methods really `transfer none`? That complicates things considerably. Right now you're potentially causing use-after-frees here (the returned pointer might just be freed here now). Maybe this should return a `&VariantType` etc?

Reply:
Hmm. I am not sure I can progress from here. If I declare `param_type` and `state_type` as returning references then I have an issue with implementing proxy action which calls another action underneath cause it already [returns an owned object](https://gtk-rs.org/gtk-rs-core/stable/latest/docs/gio/prelude/trait.ActionExt.html#method.parameter_type).

Reply:
Then this would have be handled like in e.g. `IOStreamImpl`

Reply:
Thanks! I did my best (copying from GActionGroup, he-he) but I have concerns about my implementation of `action_get_name`. It works but looks ugly.

Reply:
Ugly but looks correct except for that one comment

Reply:
Can be slightly less ugly if you use `LazyLock` instead of `OnceLock`

Reply:
My concern is rather around moving `GString` to qdata and querying it back.

Reply:
That's fine. The returned pointer after storing it in qdata is stable